public/app/plugins/datasource/prometheus/result_transformer.ts

```diff
@@ -1,234 +1,266 @@
-import _ from 'lodash';
-import TableModel from 'app/core/table_model';
-import { TimeSeries, FieldType, Labels, formatLabels, QueryResultMeta } from '@grafana/data';
-import { TemplateSrv } from 'app/features/templating/template_srv';
-
-export class ResultTransformer {
-  constructor(private templateSrv: TemplateSrv) {}
-
-  transform(response: any, options: any): Array<TableModel | TimeSeries> {
-    const prometheusResult = response.data.data.result;
-
-    if (options.format === 'table') {
-      return [
-        this.transformMetricDataToTable(
-          prometheusResult,
-          options.responseListLength,
-          options.refId,
-          options.meta,
-          options.valueWithRefId
-        ),
-      ];
-    } else if (prometheusResult && options.format === 'heatmap') {
-      let seriesList: TimeSeries[] = [];
-      for (const metricData of prometheusResult) {
-        seriesList.push(this.transformMetricData(metricData, options, options.start, options.end));
-      }
-      seriesList.sort(sortSeriesByLabel);
-      seriesList = this.transformToHistogramOverTime(seriesList);
-      return seriesList;
-    } else if (prometheusResult) {
-      const seriesList: TimeSeries[] = [];
-      for (const metricData of prometheusResult) {
-        if (response.data.data.resultType === 'matrix') {
-          seriesList.push(this.transformMetricData(metricData, options, options.start, options.end));
-        } else if (response.data.data.resultType === 'vector') {
-          seriesList.push(this.transformInstantMetricData(metricData, options));
-        }
-      }
-      return seriesList;
-    }
+import {
+  ArrayVector,
+  DataFrame,
+  Field,
+  FieldType,
+  formatLabels,
+  TIME_SERIES_TIME_FIELD_NAME,
+  TIME_SERIES_VALUE_FIELD_NAME,
+} from '@grafana/data';
+import { FetchResponse } from '@grafana/runtime';
+import templateSrv from 'app/features/templating/template_srv';
+import {
+  PromDataSuccessResponse,
+  PromMatrixData,
+  PromMetric,
+  PromValue,
+  PromVectorData,
+  TransformOptions,
+} from './types';
+
+type MatrixOrVectorResult = PromMatrixData['result'][0] | PromVectorData['result'][0];
+
+export function transform(response: FetchResponse<PromDataSuccessResponse>, options: TransformOptions) {
+  const prometheusResult = response.data.data;
```

Comment:
I see that you have changed the `prometheusResult` from `response.data.data.result` to `response.data.data` and then you are using `prometheusResult.result` and `prometheusResult.resultType`. As a personal preference, I would maybe create `prometheusResult` and `prometheusResultType` and use just those.

Reply:
This is because typescript can infer the type like this. so if resultType is scalar then it will know that the result is going to be PromScalarData.